internal/profiling/buffer.go

```diff
@@ -0,0 +1,270 @@
+
+	if atomic.LoadUint32(&q.drainingPostCheck) > 0 {
+		// Between our q load and acquired increment, a drainer began execution
+		// and switched the queues. This is NOT okay because we don't know if
+		// acquired was incremented before or after the drainer's check for
+		// acquired == writer. And we can't find which without expensive
+		// operations, which we'd like to avoid. If the acquired increment was
+		// after, we cannot write to this buffer as the drainer's collection may
+		// have already started; we must write to the other queue.
+		//
+		// Reverse our increment and retry. Since there's no SubUint32 in atomic,
+		// ^uint32(0) is used to denote -1.
+		atomic.AddUint32(&q.acquired, ^uint32(0))
+		goto reloadq
+	}
+
+	// At this point, we're definitely writing to the right queue. That is, one
+	// of the following is true:
+	//   1. No drainer is in execution.
+	//   2. A drainer is in execution and it is waiting at the acquired ==
+	//      written barrier.
+	//
+	// Let's say two Pushes A and B happen on the same queue. Say A and B are
+	// q.size apart; i.e. they get the same index. That is,
+	//
+	//   index_A = index_B
+	//   acquired_A + q.size = acquired_B
+	//
+	// We say "B has wrapped around A" when this happens. In this case, since A
+	// occurred before B, B's Push should be the final value. However, we
+	// accommodate A being the final value because wrap-arounds are extremely
+	// rare and accounting for them requires an additional counter and a
+	// significant performance penalty. Note that the below approach never leads
+	// to any data corruption.
+	index := acquired & q.mask
+	atomic.StorePointer(&q.arr[index], unsafe.Pointer(&x))
+
+	// Allows any drainWait checks to proceed.
+	atomic.AddUint32(&q.written, 1)
+}
+
+// Dereferences non-nil pointers from arr into result. Range of elements from
+// arr that are copied is [from, to). Assumes that the result slice is already
+// allocated and is large enough to hold all the elements that might be copied.
+// Also assumes mutual exclusion on the array of pointers.
+func dereferenceAppend(result []interface{}, arr []unsafe.Pointer, from, to uint32) []interface{} {
+	for i := from; i < to; i++ {
+		// We have mutual exclusion on arr, there's no need for atomics.
+		x := (*interface{})(arr[i])
+		if x != nil {
+			result = append(result, *x)
+		}
+	}
+	return result
+}
+
+// Allocates and returns an array of things Pushed in to the circular buffer.
+// Push order is not maintained; that is, if B was Pushed after A, drain may
+// return B at a lower index than A in the returned array.
+func (cb *circularBuffer) Drain() []interface{} {
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+
+	qs := make([]*queue, 0)
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		qs = append(qs, cb.qp[i].switchQueues())
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(int(numCircularBufferPairs))
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		go func(qi uint32) {
+			qs[qi].drainWait()
```

Comment:
Can this happen?

| Push                                | Drain                                      |
|-------------------------------------|--------------------------------------------|
| load q                              |                                            |
|                                     | drainWait(), returns                       |
| q.acquired++                        |                                            |
| q.drainingPostCheck == 0, not retry |                                            |
|                                     | q.drainingPostCheck = 1 (this is too late) |

Push will write to this queue, while Drain is draining it, right?

Reply:
Excellent catch! I've changed the algorithm slightly to do the barrier check before the switching happens.